src/Security/Authorization/Core/src/AuthorizationHandlerContext.cs

```diff
@@ -59,6 +60,11 @@ public class AuthorizationHandlerContext
         /// </summary>
         public virtual IEnumerable<IAuthorizationRequirement> PendingRequirements { get { return _pendingRequirements; } }
 
+        /// <summary>
+        /// Gets the reasons why authorization has failed.
+        /// </summary>
+        public virtual IEnumerable<AuthorizationFailureReason> FailureReasons { get { return _failedReasons ?? Array.Empty<AuthorizationFailureReason>(); } }
```

Comment:
I'm not sure which one is better, but in case of `Array.Empty` if user store the result of `FailureReasons` and the call `Fail` then the result would still be empty.
```suggestion
        public virtual IEnumerable<AuthorizationFailureReason> FailureReasons { get { return _failedReasons ??= new List<AuthorizationFailureReason>(); } }
```

Reply:
I think the scenario where there are no reasons is more common, let's stick with Array.Empty.

Reply:
```suggestion
        public virtual IEnumerable<AuthorizationFailureReason> FailureReasons { get { return _failedReasons != null ? _failedReasons : Array.Empty<AuthorizationFailureReason>(); } }
```